### ISSUE TITLE
collectd-5.8.0.rc0: be16toh / be64toh not supported on Solaris 11.3

### ISSUE THREAD

**Issue by skrueger8** [collectd/collectd#2523]
*   Version of collectd: collectd-5.8.0.rc0
*   Operating system / distribution: Solaris 11.3 SPARC

## Expected behavior

collectd 5.8 compiles fine

## Actual behavior

it fails with:

```
...
libtool: link: /opt/developerstudio12.6/bin/cc -Wall -m64 -xO4 -xarch=sparcvis2 -Qoption cg -xregs=no%appl -W2,-xwrap_int -xc99=all -mt -KPIC -DPIC -xmemalign=16s -m64 -M /usr/lib/ld/map.noexbss -M /usr/lib/ld/map.pagealign -Bdirect -zignore -zstrip-class=comment -zaslr=enable -znxstack=enable -znxheap=enable -ztype=pie -o .libs/collectd-nagios src/collectd_nagios-collectd-nagios.o  ./.libs/libcollectdclient.so -ladimalloc -lm -lgcrypt -lgpg-error -lsocket -mt -R/opt/collectd/collectd57/lib
Undefined                       first referenced
 symbol                             in file
be16toh                             ./.libs/libcollectdclient.so
be64toh                             ./.libs/libcollectdclient.so
ld: fatal: symbol referencing errors
```

## Steps to reproduce

*  `./configure CC=cc CFLAGS='-m64 -xO4 -xarch=sparcvis2 -Qoption cg -xregs=no%appl -W2,-xwrap_int -xc99=all -mt -KPIC -DPIC -xmemalign=16s' LDFLAGS='-m64 -M /usr/lib/ld/map.noexbss -M /usr/lib/ld/map.pagealign -Bdirect -zignore -zstrip-class=comment -zaslr=enable -znxstack=enable -znxheap=enable -ztype=pie' --prefix=... --disable-silent-rules --disable-werror --enable-snmp=no`
* fix some errors which are already fixed in Git...
* `gmake`

**Comment by trustchk** [comment]
This seems to be the same issue I tried to fix in PR #2517 for AIX/GCC, from what I could find `sparcvis2` is a SPARC V9 and to my understanding should not need to do any conversion from big endian to host byte order.
A quick search didn't turn up much useful information on that matter, so in order to patch this one would need to differentiate Solaris on SPARC (big endian) and Solaris on x86 (little endian). Untested braindump for `src/libcollectdclient/network_parse.c` :
```C
#ifdef __sun
#ifdef __sparc
#define be16toh(x) (x)
#define be64toh(x) (x)
#else /* -> x86 */
#define be16toh(x) __uint16_identity(x)
#define be64toh(x) __uint64_identity(x)
#endif
#endif /* sun */
```

**Comment by octo** [comment]
*be16toh()* may be replaced with the more portable *ntohs()*.

The glibc uses some preprocessor magic like this:

```c
#if __BYTE_ORDER == __BIG_ENDIAN
```

Do Solaris and AIX have something similar we could do? I'd prefer not to make too many assumptions about architectures, such as "if it's not sparc, it must be x86" …

Best regards,
—octo

**Comment by trustchk** [comment]
well, in the above case, `__sun` limits it to Solaris, there are only two choices available where it can run: SPARC and x86 :)
Actually, the problem is that for Solaris and AIX, there's the makers own build toolchain (sun studio(?) and xlC) that define some things a little different, and GCC. I only have access to GCC on AIX. Also, there is no glibc on these machines as the vendors supply their own libc, some of the headers might contain some surprises. I remember running into one when porting another project to AIX.

AIX has `/usr/include/sys/machine.h`, excerpt:
```C
/*
 * Definitions for byte order,
 * according to byte significance from low address to high.
 */
#define LITTLE_ENDIAN   1234    /* least-significant byte first (vax)         */
#define BIG_ENDIAN      4321    /* most-significant byte first (IBM, net)     */

#define BYTE_ORDER      BIG_ENDIAN
```
There seems to be no `__BYTE_ORDER`, this looks like a glibc-thing, but that can be defined later based on the above info. As for PowerPC, the CPU can run in little endian mode for things like Linux on Power, but I don't have that around to look how things are defined there. AIX is big endian only.

AIX defines `ntohs` in `arpa/inet.h` just like on linux. Basically, AIX is strictly POSIX. I assume the same for Solaris, though I cannot easily confirm that. Anyway, ntohs and the like look like a very good solution and I'd be happy to run some tests on AIX.

**Comment by skrueger8** [comment]
Seems like Developer Studio 12.6 defines `__BYTE_ORDER__`

```
$ cc -xdumpmacros -E /dev/null 2>&1 | grep ORDER
#define __ORDER_LITTLE_ENDIAN__ 1234
#define __ORDER_BIG_ENDIAN__ 4321
#define __BYTE_ORDER__ __ORDER_BIG_ENDIAN__
```

Using `__BYTE_ORDER ` seems to work, too.

```
$ cat macro.c
#include <stdio.h>

int
main(void)
{

#if __BYTE_ORDER == __LITTLE_ENDIAN
    printf("__BYTE_ORDER == _LITTLE_ENDIAN\n");
#endif

        return 0;
}
```

```
$ cc -m64 macro.c
$ ./a.out
__BYTE_ORDER == _LITTLE_ENDIAN
```

**Comment by octo** [comment]
I think this was fixed with #2517.

**Comment by skrueger8** [comment]
htobe64 is still used in src/libcollectdclient/network_parse_test.c

```
$ gmake check
...
libtool: link: /opt/developerstudio12.6/bin/cc -Wall -m64 -xO4 -xarch=sparcvis2 -xc99=all -features=extensions -KPIC -DPIC -W2,-xwrap_int -Qoption cg -xregs=no%appl -xmemalign=16s -mt -m64 -M /usr/lib/ld/map.noexbss -M /usr/lib/ld/map.pagealign -Bdirect -zignore -zstrip-class=comment -zaslr=enable -znxstack=enable -znxheap=enable -ztype=pie -o test_libcollectd_network_parse src/libcollectdclient/test_libcollectd_network_parse-network_parse_test.o  -lgcrypt -lgpg-error -mt
Undefined                       first referenced
 symbol                             in file
htobe64                             src/libcollectdclient/test_libcollectd_network_parse-network_parse_test.o
ld: fatal: symbol referencing errors
gmake[2]: *** [test_libcollectd_network_parse] Error 2
```

**Comment by rubenk** [comment]
Indeed, reopening issue.

**Comment by rubenk** [comment]
Should be fixed by 720fdc32913bfa5413716beccf54dcd9a3d7b506. @skrueger8 could you give this a try?

**Comment by skrueger8** [comment]
Compiles fine on Solaris 11.4 Beta / SPARC. Good work everyone!

```
============================================================================
Testsuite summary for collectd 5.8.0.184.g34471c0
============================================================================
# TOTAL: 12
# PASS:  12
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```
